src/client/gui/lib/vm_details/terminal.dart

```diff
@@ -1,102 +1,189 @@
+    if (!running) return null;
 
+    final grpcClient = ref.read(grpcClientProvider);
+    final sshInfo = await grpcClient.sshInfo(arg.vmName).onError((err, stack) {
+      ref
+          .notifyError((error) => 'Failed to get SSH information: $err')
+          .call(err, stack);
+      return null;
+    });
+    if (sshInfo == null) return null;
+
+    final terminal = Terminal(maxLines: 10000);
     final receiver = ReceivePort();
     final errorReceiver = ReceivePort();
     final exitReceiver = ReceivePort();
-    isolate = await Isolate.spawn(
-      sshIsolate,
-      SshShellInfo(
-        sender: receiver.sendPort,
-        width: thisTerminal.viewWidth,
-        height: thisTerminal.viewHeight,
-        sshInfo: sshInfo,
-      ),
-      onError: errorReceiver.sendPort,
-      onExit: exitReceiver.sendPort,
-      errorsAreFatal: true,
-    );
-
-    vmShellsNotifier.update((state) => state + 1);
 
     errorReceiver.listen((es) {
+      final (Object? error, String? stack) = (es[0], es[1]);
       logger.e(
-        'Error from ${widget.name} ssh isolate',
-        error: es[0],
-        stackTrace: es[1] != null ? StackTrace.fromString(es[1]) : null,
+        'Error from $arg ssh isolate',
+        error: error,
+        stackTrace: stack != null ? StackTrace.fromString(stack) : null,
       );
     });
+
     exitReceiver.listen((_) {
-      logger.d('Exited ${widget.name} ssh isolate');
+      logger.d('Exited $arg ssh isolate');
       receiver.close();
       errorReceiver.close();
       exitReceiver.close();
-      vmShellsNotifier.update((state) => max(0, state - 1));
-      if (mounted) setState(() => terminal = null);
+      stop();
     });
+
     receiver.listen((event) {
       switch (event) {
         case final SendPort sender:
-          thisTerminal.onOutput = sender.send;
-          thisTerminal.onResize = (w, h, pw, ph) => sender.send([w, h, pw, ph]);
+          terminal.onOutput = sender.send;
+          terminal.onResize = (w, h, pw, ph) => sender.send([w, h, pw, ph]);
         case final String data:
-          thisTerminal.write(data);
+          terminal.write(data);
         case null:
-          logger.i('Ssh session for ${widget.name} has exited');
-          isolate?.kill(priority: Isolate.immediate);
+          stop();
       }
     });
+
+    isolate = await Isolate.spawn(
+      sshIsolate,
+      SshShellInfo(
+        sender: receiver.sendPort,
+        width: terminal.viewWidth,
+        height: terminal.viewHeight,
+        sshInfo: sshInfo,
+      ),
+      onError: errorReceiver.sendPort,
+      onExit: exitReceiver.sendPort,
+      errorsAreFatal: true,
+    );
+
+    ref.read(runningShellsProvider(arg.vmName).notifier).update((state) {
+      return state + 1;
+    });
+    return terminal;
+  }
+
+  Future<void> start() async {
+    state = await lock.synchronized(_initShell);
+  }
+
+  void stop() {
+    _dispose();
+    state = null;
+  }
+
+  void _dispose() {
```

Comment:
Why is it that when I close the last open terminal of a running instance by clicking the X, it opens a new one, but if I input Ctrl+D or `exit`, it closes it and brings up the "open shell" view? Shouldn't we allow closing all terminals and default to the "open shell" all the time?

Reply:
See the explanation I gave above about the current state of shell autostart behavior.
But still, I really agree that it's weird you can't close the last shell without automatically opening a new one. So I added that behavior. Let me know if it is more intuitive now.

Reply:
Yes, I think that is better. However, I just tried it and the "open shell" button has no effect when the last tab was closed with the X.

Reply:
Yep, indeed that is the case. Will fix.